bee-message/src/payload/mod.rs

```diff
@@ -0,0 +1,352 @@
+        match self {
+            Self::Data(e) => write!(f, "Error unpacking data payload: {}", e),
+            Self::Dkg(e) => write!(f, "Error unpacking DKG payload: {}", e),
+            Self::Fpc(e) => write!(f, "Error unpacking FPC payload: {}.", e),
+            Self::Indexation(e) => write!(f, "Error unpacking indexation payload: {}.", e),
+            Self::InvalidKind(kind) => write!(f, "Invalid payload kind: {}.", kind),
+            Self::SaltDeclaration(e) => write!(f, "Error unpacking salt declaration payload: {}", e),
+            Self::Transaction(e) => write!(f, "Error unpacking transaction payload: {}", e),
+        }
+    }
+}
+
+impl From<DataUnpackError> for PayloadUnpackError {
+    fn from(error: DataUnpackError) -> Self {
+        Self::Data(error)
+    }
+}
+
+impl From<DkgUnpackError> for PayloadUnpackError {
+    fn from(error: DkgUnpackError) -> Self {
+        Self::Dkg(error)
+    }
+}
+
+impl From<FpcUnpackError> for PayloadUnpackError {
+    fn from(error: FpcUnpackError) -> Self {
+        Self::Fpc(error)
+    }
+}
+
+impl From<IndexationUnpackError> for PayloadUnpackError {
+    fn from(error: IndexationUnpackError) -> Self {
+        Self::Indexation(error)
+    }
+}
+
+impl From<SaltDeclarationUnpackError> for PayloadUnpackError {
+    fn from(error: SaltDeclarationUnpackError) -> Self {
+        Self::SaltDeclaration(error)
+    }
+}
+
+impl From<TransactionUnpackError> for PayloadUnpackError {
+    fn from(error: TransactionUnpackError) -> Self {
+        Self::Transaction(error)
+    }
+}
+
+impl From<UnknownTagError<u32>> for PayloadUnpackError {
+    fn from(error: UnknownTagError<u32>) -> Self {
+        match error {
+            UnknownTagError(tag) => Self::InvalidKind(tag),
+        }
+    }
+}
+
+impl From<Infallible> for PayloadUnpackError {
+    fn from(error: Infallible) -> Self {
+        match error {}
+    }
+}
+
+/// A generic payload that can represent different types defining message payloads.
+#[non_exhaustive]
+#[derive(Clone, Debug, Eq, PartialEq)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Serialize, serde::Deserialize),
+    serde(tag = "type", content = "data")
+)]
+pub enum Payload {
+    /// A dRNG application message payload.
+    ApplicationMessage(Box<ApplicationMessagePayload>),
+    /// A dRNG beacon payload.
+    Beacon(Box<BeaconPayload>),
+    /// A dRNG collective beacon payload.
+    CollectiveBeacon(Box<CollectiveBeaconPayload>),
+    /// A pure data payload.
+    Data(Box<DataPayload>),
+    /// A dRNG DKG payload.
+    Dkg(Box<DkgPayload>),
+    /// An FPC payload.
+    Fpc(Box<FpcPayload>),
+    /// An indexation payload.
+    Indexation(Box<IndexationPayload>),
+    /// A salt declaration payload.
+    SaltDeclaration(Box<SaltDeclarationPayload>),
+    /// A transaction payload.
+    Transaction(Box<TransactionPayload>),
+}
+
+impl Payload {
+    /// Returns the payload kind of a `Payload`.
+    pub fn kind(&self) -> u32 {
```

Comment:
Random ~~bad~~ idea: maybe having a kind should be some sort of trait.

Reply:
I was thinking something similar with payload versions as well...I'll think some more on it :trollface:

Reply:
Done now 👍